### PR TITLE
LABSCon 2022 Techniques

### PR DESCRIPTION
**Pull request by moohax**
New PR for CLA. 

Changes:
---------------------
tf -> tensorflow
np -> numpy

### CODE REVIEW

File: python/tensorflow-load-op-library.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import tensorflow
+
+path = ""
+
+#rule-id
```
Comment (GrosQuildu) [inline python/tensorflow-load-op-library.py:5]:
```suggestion
# rule-id: tensorflow-load-op-library
```

File: python/tensorflow-load-op-library.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+rules:
+- id: tensorflow-load-op-library
+  patterns:
+    - pattern-either: 
+      - pattern: |
```
Comment (GrosQuildu) [inline python/tensorflow-load-op-library.yml:5]: Incorrect syntax - crashing semgrep

**Review by GrosQuildu — CHANGES_REQUESTED** [review]
I added metadata to few rules. All rules need similar fields. (We will make CONTRIBUTING file soon with all requirements for new rules).

Left a few comments - some rules can be merged; some have invalid tests.

All tests need improvements:
* add `# ruleid: <name of rule>` where the semgrep should find the issue
* add `# ok: <name of rule>` where semgrep should not find issue

I am  a bit worried about number of false positives - is using all these methods always insecure, or can we filter-out some common false positives (e.g., using hardcoded paths)?

If possible, we can enhance messages with guidance on how to fix the issues - propose more secure alternatives or e.g., sanitizing inputs.

Rules like these can be simplified:
```semgrep
patterns:
    - pattern-either: 
      - pattern: |
      tensorflow.load_op_library($PATH)
```

with

```semgrep
patterns:
  - pattern: |
    tensorflow.load_op_library($PATH)
```

Once everything is fixed, run `semgrep --test ./`

File: python/numpy-f2py-compile.yml [hunk 3]
```diff
@@ -1,11 +1,19 @@
...
+  message: Compiling arbitrary code can result in code execution. Ensure the source code is from a trusted location
   languages: [python]
   severity: WARNING
+  metadata:
+    category: security
+    cwe: "CWE-676: Use of Potentially Dangerous Function"
+    subcategory: [audit]
+    confidence: LOW
+    likelihood: LOW
+    impact: MEDIUM
```
Comment (moohax) [inline python/numpy-f2py-compile.yml:12]: Arbitrary code execution is medium?

File: python/numpy-in-pytorch-modules.py [hunk 4]
```diff
@@ -1,11 +1,16 @@
 import torch.nn as nn
-import numpy
+import numpy as np
```
Comment (moohax) [inline python/numpy-in-pytorch-modules.py:2]: I was told in the other PR that `import numpy` was preferred.

File: python/onnx-convert-ort.py [hunk 5]
```diff
@@ -0,0 +1,2 @@
+import os
+os.system("python -m onnxruntime.tools.convert_onnx_models_to_ort ./bin/onnx --custom_op_library ./bin/custom_op.dll")
```
[the remark this answers is not in the document]
Comment (moohax) [reply python/onnx-convert-ort.py:4]: Not terribly familiar with semgrep - can I assign this to you?

File: python/onnx-session-options.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+rules:
+- id: onnx-session-options
```
[the remark this answers is not in the document]
Comment (moohax) [reply python/onnx-session-options.yml:2]: There are two techniques - one is via command line and the other via configuration options. That is, you might allow users to provide some config file that gets passed in at runtime.

File: python/numpy-in-pytorch-modules.py (hunk 4 above)
Comment (GrosQuildu) [reply python/numpy-in-pytorch-modules.py:2]: https://github.com/trailofbits/semgrep-rules/pull/12#discussion_r1047158663 was about `numpy` in semgrep rule, not in python test.
That is, in semgrep we want to catch all `numpy` package uses, [even if they are aliased/renamed to something like `np`](https://semgrep.dev/docs/writing-rules/pattern-syntax/#imports).

On the other hand, in tests we want to verify our semgrep rule with not-straightforward code, like e.g., we can `import numpy as something`.

File: python/numpy-f2py-compile.yml (hunk 3 above)
Comment (GrosQuildu) [reply python/numpy-f2py-compile.yml:12]: My mistake, in similar rules we set HIGH. Feel free to change ;)

File: python/onnx-session-options.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (GrosQuildu) [reply python/onnx-session-options.yml:2]: semgrep cannot detect that

**Comment by GrosQuildu** [comment]
To limit false positive, removed detection of hardcoded strings. This is in accordance with similar, official semgrep rules.

Removed:
 * onnx-convert-ort - semgrep doesn't support bash language yet
 * pandas-read-* - reading json/csv is not inherently insecure, it depends form where you get data. There are advanced rules that can track data flow and determine if reading specific file is safe. Consider extending such rule with pandas. I guess CodeQL should be a good match here.
* pickle-load - already exists` r/python.lang.security.deserialization.pickle.avoid-pickle`
* tensorflow-load-op-library - merged with tensorflow-load-op-library

**Review by GrosQuildu — APPROVED** [review]

**Comment by GrosQuildu** [comment]
Thanks a lot @moohax ! Finally merged the rules.

**Pull request merged**